File: src/normalize/summarize.ts
import type { SearchResult } from "../providers/types.js";
import { loadConfig } from "../config.js";
import { chatCopilot } from "../ai/copilot.js";
import { chatLocal } from "../ai/local.js";

export async function summarizeResults(
  query: string,
  results: SearchResult[],
): Promise<string | undefined> {
  const top = results.slice(0, 5);
  if (top.length === 0) return undefined;

  const context = top
    .map((r, i) => `${i + 1}. [${r.title}](${r.url})\n   ${r.snippet}`)
    .join("\n\n");

  const prompt = `Summarize the following search results for the query "${query}". Provide a concise 2-3 sentence overview of what the results indicate. Do not include any URLs or links in the summary.\n\nResults:\n${context}`;

  try {
    return await chatWithAI(prompt);
  } catch {
    // AI unavailable — return no summary rather than failing
    return undefined;
  }
}

async function chatWithAI(prompt: string): Promise<string> {
  const config = loadConfig();

  // Try Copilot first
  if (config.copilotToken) {
    try {
      return await chatCopilot(prompt, config.copilotToken);
    } catch {
      // Fall through to local model
    }
  }

  // Try local model
  if (config.localModelUrl) {
    return await chatLocal(prompt, config.localModelUrl, config.localModelName);
  }

  throw new Error("No AI provider configured");
}

File: src/evoscry/normalize/summarize.py
"""AI-powered summarization of search results."""

from __future__ import annotations

from evoscry.ai.copilot import chat_copilot
from evoscry.ai.local import chat_local
from evoscry.config import load_config


async def summarize_results(query: str, results: list[dict]) -> str | None:
    """Summarize top search results using AI. Returns None if AI unavailable."""
    top = results[:5]
    if not top:
        return None

    context = "\n\n".join(
        f"{i+1}. [{r['title']}]({r['url']})\n   {r.get('snippet', '')}"
        for i, r in enumerate(top)
    )
    prompt = (
        f'Summarize the following search results for the query "{query}". '
        "Provide a concise 2-3 sentence overview of what the results indicate. "
        "Do not include any URLs or links in the summary.\n\n"
        f"Results:\n{context}"
    )

    try:
        return await _chat_with_ai(prompt)
    except Exception:
        return None


async def _chat_with_ai(prompt: str) -> str:
    config = load_config()

    # Try Copilot first
    if config.copilot_token or config.copilot_refresh_token:
        try:
            return await chat_copilot(
                prompt,
                config.copilot_token,
                config.copilot_refresh_token,
            )
        except Exception:
            pass

    # Try local model
    if config.local_model_url:
        return await chat_local(prompt, config.local_model_url, config.local_model_name)

    raise RuntimeError("No AI provider configured")
